vignettes/sass.Rmd

```diff
@@ -1,165 +1,402 @@
+
+For some examples, please see the [Sass mixin documentation](https://sass-lang.com/documentation/at-rules/mixin).
+
+### More basics
+
+This vignette intentionally doesn't try to re-invent the existing and wonderful [Sass documentation](https://sass-lang.com/documentation). There you'll find many more useful things as you start to write more Sass, such as [control flow](https://sass-lang.com/documentation/at-rules/control), [lists](https://sass-lang.com/documentation/values/lists), [maps](https://sass-lang.com/documentation/values/maps), [interpolation](https://sass-lang.com/documentation/interpolation), and more.
+
+## Composable sass {#layering}
+
+To make Sass code more composable with other Sass code (e.g., allowing others to change your variable defaults or import a function or mixin you've defined), consider partitioning your Sass code into a `sass::sass_layer()`. The main idea is to split your Sass into 3 parts: `defaults` (i.e. variable defaults), `declarations` (i.e., [functions](#functions), [mixins](#mixins), etc), and `rules` (i.e., styling rules).
+
+```{r}
+layer1 <- sass_layer(
+  defaults = list("body-bg" = "white !default"),
+  declarations = sass_file("color-contrast.scss"),
+  rules = "body{background-color: $body-bg; color: color-contrast($body-bg)}"
+)
+sass(layer1)
 ```
 
-```{r echo = FALSE}
-knitr::include_app("https://gallery.shinyapps.io/sass-size/", height = "600px")
+This allows downstream `sass_layer()`s to be `sass_layer_merge()`d into a single layer, where those downstream layers are granted higher priority. More specifically, this means: 
+
+* `defaults` for `layer2` are placed _before_ `defaults` for `layer1`.
+  * Allowing downstream Sass to override variable defaults in upstream Sass.
+* `declarations` for `layer2` are placed _after_ `declarations` for `layer1`.
+  * Allowing downstream Sass to use functions and mixins defined in upstream Sass.
+* `rules` for `layer2` are placed _after_ `rules` for `layer1`.
+  * Allowing downstream Sass rules to override upstream rules.
+
+```{r}
+layer2 <- sass_layer(
+  defaults = list("body-bg" = "white !default")
+)
+sass(sass_layer_merge(layer1, layer2))
 ```
 
+### Resolving relative imports {#imports-relative}
 
-## CSS Formatting
+Another problem that `sass_layer()` helps solve is that sometimes your Sass code might want to [importing](#imports) a local file using a relative path that _you_ know how to resolve, but not necessarily the person who eventually compiles your Sass. To solve this issue, provide a named character vector to `file_attachments`, pointing the relevant relative path(s) to the appropriate absolute path(s). Here's a contrived example of how that might look ([here's](https://github.com/rstudio/bootstraplib/blob/9c973e9/R/layers.R#L75-L77) a more real example of using it in an R package).
```

Comment:
```suggestion
Another problem that `sass_layer()` helps solve is that sometimes your Sass code might want to [import](#imports) a local file using a relative path that _you_ know how to resolve, but not necessarily the person who eventually compiles your Sass. To solve this issue, provide a named character vector to `file_attachments`, pointing the relevant relative path(s) to the appropriate absolute path(s). Here's a contrived example of how that might look ([here's](https://github.com/rstudio/bootstraplib/blob/9c973e9/R/layers.R#L75-L77) a more real example of using it in an R package).
```

Reply:
(That suggestion was just changing "importing" to "import")

Reply:
Also maybe should make it clear that "local file" in this case generally means an image or font file, and definitely not an .scss file. And `@import` in this case isn't in the Sass sense of the word, but in the regular CSS usage.